src/services/formatting/rules.ts

```diff
@@ -649,6 +662,28 @@ module ts.formatting {
             return context.TokensAreOnSameLine();
         }
 
+        static IsEndOfDecoratorContextOnSameLine(context: FormattingContext): boolean {
+            return context.TokensAreOnSameLine() &&
+                context.contextNode.decorators &&
+                Rules.NodeIsInDecoratorContext(context.currentTokenParent) &&
+                !Rules.NodeIsInDecoratorContext(context.nextTokenParent);
+        }
+
+        static NodeIsInDecoratorContext(node: Node): boolean {
+            if (node.parserContextFlags & ParserContextFlags.Decorator) {
+                return true;
+            }
+            while (node) {
+                if (isExpression(node)) {

```

Comment:
can you just walk up until you find a decorator, or just use hasAncestor/getAncestor